# GRG Paper Experiments

Summary of the commands/arguments used for experiments from the [GRG paper](https://www.nature.com/articles/s43588-024-00739-9).
Tagged version [v1.01-paper](https://github.com/aprilweilab/grgl/releases/tag/v1.01-paper) of GRGL is the most recent version
used for experiments.

XSI and Savvy commands/APIs that we compared against are specified in the [GRG paper supplement](https://www.biorxiv.org/content/10.1101/2024.04.23.590800v2.supplementary-material).

GRG construction can be particularly sensitive to the chosen `-p` and `-t` parameters, see the paper for details.

We used SSD-based storage for testing whenever possible, since it helps the non-GRG methods perform faster (they
tend to be more IO-bound). The [Docker container defined in this repository](https://github.com/aprilweilab/grgl/blob/main/Dockerfile)
was used as the basis for our UKB experiments, since they had to be done in the cloud.

## UKB GRG Construction

IGD conversion is the first step in GRG construction, because we need a faster input format than
.vcf.gz. Conversion is the same for all chromosomes, and looks like this:
```
gconvertlz4 ukb20279_c3_b0_v1.vcf.gz ukb20279_c3_b0_v1.vcf.igd
```
where `gconvertlz4` is an alias for [this script](https://github.com/aprilweilab/grgl/blob/main/scripts/convertlz4.sh)
which uses [gconverter](https://github.com/aprilweilab/grgl/blob/main/src/gconverter.cpp) to do the
IGD conversion, and then lz4 compresses the result for better cloud performance.

If you want to do IGD conversion, you can either use `grg convert` (which calls `gconverter` under the hood) which is part of GRGL, or you can use `igdtools` which is part of [picovcf](https://github.com/aprilweilab/picovcf).

Each UKB chromosome had (potentially) different `-p` and `-t` parameters, so here is the full list:
```
grg construct -p 350 -t 3 -j 70 ukb20279_c1_b0_v1.v3.igd.lz4
grg construct -p 350 -t 3 -j 70 ukb20279_c2_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c3_b0_v1.v3.igd.lz4
grg construct -p 280 -t 4 -j 70 ukb20279_c4_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c5_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c6_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c7_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c8_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c9_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c10_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c11_b0_v1.v3.igd.lz4
grg construct -p 240 -t 4 -j 70 ukb20279_c12_b0_v1.v3.igd.lz4
grg construct -p 210 -t 3 -j 70 ukb20279_c13_b0_v1.v3.igd.lz4
grg construct -p 240 -t 3 -j 70 ukb20279_c14_b0_v1.v3.igd.lz4
grg construct -p 240 -t 3 -j 70 ukb20279_c15_b0_v1.v3.igd.lz4
grg construct -p 240 -t 3 -j 70 ukb20279_c16_b0_v1.v3.igd.lz4
grg construct -p 180 -t 3 -j 70 ukb20279_c17_b0_v1.v3.igd.lz4
grg construct -p 180 -t 3 -j 70 ukb20279_c18_b0_v1.v3.igd.lz4
grg construct -p 140 -t 3 -j 70 ukb20279_c19_b0_v1.v3.igd.lz4
grg construct -p 140 -t 3 -j 70 ukb20279_c20_b0_v1.v3.igd.lz4
grg construct -p 70 -t 4 -j 70 ukb20279_c21_b0_v1.v3.igd.lz4
grg construct -p 70 -t 4 -j 70 ukb20279_c22_b0_v1.v3.igd.lz4
```

## UKB GWAS, Dot Product, and Allele Frequency

The "repeat tests" were done with a modified version of GRGL to let us specify
how many times to run an experiment. This isn't generally useful functionality, so
it was not released in the code.

`grgp` is the C++ executable that performs all the functionality of the more
first-class command `grg process` (which is just a Python wrapper around `grgp`).

The dot-product commands were, e.g.:
```
grgp -a -b ukb20279_c22_b0_v1.v3.igd.final.grg > dot_prod.grg.chr22.stdout.txt
```

The GWAS in UKB was done using a random phenotype, which is generated by `grgp` when
no phenotype is specified (which would be via `-p` or `--phenotype` followed by a file).
```
grgp -a ukb20279_c22_b0_v1.v3.igd.final.grg > gwas.grg.chr22.stdout.txt
```

Allele frequency was calculated via:
```
grgp --freq ukb20279_c22_b0_v1.v3.igd.final.grg > full_af.grg.chr22.stdout.txt
```


## 1,000 Genomes GRG Construction

All chromosomes were build using the same parameters. Here is chromosome 10 as an example:
```
grg construct -p 50 -j 20 -t 32 ALL.chr10.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.v3.igd
```

## Simulation GWAS

Our simulation tests used simulated phenotype input files, instead of just random ones, so
the command line is slightly different. Example:

```
grgp -a -p simulation-source-10-100000000.plink.ncausal.1000.hsq.0.01.BASELINE.42.phen simulation-source-10-100000000.vcf.v3.igd.final.grg > full_gwas.10.out
```
