{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97b7771a-1daf-4ec0-a26b-4e169e4332d6",
   "metadata": {},
   "source": [
    "# Create IGDs and Converting IGD to GRG\n",
    "\n",
    "This tutorial covers two topics:\n",
    "1. Creating an IGD file from a `.vcf.gz`.\n",
    "2. Converting an IGD file to a GRG\n",
    "\n",
    "IGD files store referenced-aligned genotype data in a very simple sparse matrix format, without compression. An IGD file is often smaller than the corresponding `.vcf.gz` file, and is almost always faster (usually `15x` or more) to access. See [the paper](https://pmc.ncbi.nlm.nih.gov/articles/PMC11838554/) if you're interested in comparison experiments.\n",
    "\n",
    "Many datasets are stored in `.vcf.gz` format by \"default.\" If these datasets are large, they are usually stored using [BGZIP](https://www.htslib.org/doc/bgzip.html) so that they can be indexed for semi-random access. The two different kinds of index files for `BGZIP` are [tabix](https://www.htslib.org/doc/tabix.html) or [bcftools](https://samtools.github.io/bcftools/bcftools.html). We only support `tabix`-style indexes.\n",
    "\n",
    "**What you'll need:**\n",
    "* Python dependencies \"pygrgl\" and \"igdtools\": `pip install pygrgl igdtools`\n",
    "* Command line tool \"wget\" and \"tabix\": `sudo apt install wget tabix` (or your distribution's equivalent)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "210ae4f7-d602-49db-9e0a-2ae1291bc04f",
   "metadata": {},
   "source": [
    "### Get Dataset\n",
    "\n",
    "For our example, we'll just download a very small simulated dataset that is stored as `.vcf.gz`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc0fa2a0-1b40-4f38-9293-0cb7a9076c4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2026-02-06 12:39:40--  https://github.com/aprilweilab/grg_pheno_sim/raw/refs/heads/main/demos/data/test-200-samples.vcf.gz\n",
      "Resolving github.com (github.com)... 140.82.113.4\n",
      "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://raw.githubusercontent.com/aprilweilab/grg_pheno_sim/refs/heads/main/demos/data/test-200-samples.vcf.gz [following]\n",
      "--2026-02-06 12:39:40--  https://raw.githubusercontent.com/aprilweilab/grg_pheno_sim/refs/heads/main/demos/data/test-200-samples.vcf.gz\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.109.133, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 494022 (482K) [application/octet-stream]\n",
      "Saving to: ‘igd_convert.example.vcf.gz’\n",
      "\n",
      "     0K .......... .......... .......... .......... .......... 10% 1.95M 0s\n",
      "    50K .......... .......... .......... .......... .......... 20% 2.34M 0s\n",
      "   100K .......... .......... .......... .......... .......... 31% 42.2M 0s\n",
      "   150K .......... .......... .......... .......... .......... 41% 4.13M 0s\n",
      "   200K .......... .......... .......... .......... .......... 51% 5.84M 0s\n",
      "   250K .......... .......... .......... .......... .......... 62% 4.09M 0s\n",
      "   300K .......... .......... .......... .......... .......... 72% 3.48M 0s\n",
      "   350K .......... .......... .......... .......... .......... 82% 4.20M 0s\n",
      "   400K .......... .......... .......... .......... .......... 93% 8.09M 0s\n",
      "   450K .......... .......... .......... ..                   100% 7.34M=0.1s\n",
      "\n",
      "2026-02-06 12:39:41 (4.09 MB/s) - ‘igd_convert.example.vcf.gz’ saved [494022/494022]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Download a small example dataset\n",
    "wget https://github.com/aprilweilab/grg_pheno_sim/raw/refs/heads/main/demos/data/test-200-samples.vcf.gz -O igd_convert.example.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e5c539-53a1-4728-9c82-50dea2266a30",
   "metadata": {},
   "source": [
    "## Convert to IGD\n",
    "\n",
    "If we want to convert from `.vcf.gz` to IGD with a single thread, then we do not need a `tabix` index. However, if we want to use multiple threads (i.e., `-j 2` below) then a tabix index will provide a performance improvement. So here we first `tabix` index the dataset, and then convert to IGD with `igdtools`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0fd638af-a110-4721-9499-27f12b312144",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 5447 total variants\n",
      "Of which 3058 were written sparsely\n",
      "Wrote 5446 total variants\n",
      "Of which 3170 were written sparsely\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Index the file. Note: usually your dataset will come with an index already.\n",
    "tabix igd_convert.example.vcf.gz\n",
    "\n",
    "# -j controls how many threads to use.\n",
    "igdtools -j 2 igd_convert.example.vcf.gz -o igd_convert.example.igd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fd474ae-ba5a-46fe-956e-413f2ee2138e",
   "metadata": {},
   "source": [
    "## Convert to GRG\n",
    "\n",
    "If we have an IGD file (either because we converted it, or that is how our dataset came) then we can construct a GRG easily:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f790a4fb-f270-4b8f-a1d7-8d62b55a78b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing input file in 85 parts.\n",
      "Auto-calculating number of trees per part.\n",
      "Converting segments of input data to graphs\n",
      "100%|██████████| 85/85 [00:00<00:00, 127.28it/s]\n",
      "Merging...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== GRG Statistics ===\n",
      "Nodes: 15481\n",
      "Edges: 93351\n",
      "Samples: 400\n",
      "Mutations: 10893\n",
      "Ploidy: 2\n",
      "Phased: true\n",
      "Populations: 0\n",
      "Range of mutations: 55829 - 9999127\n",
      "Specified range: 0 - 10894\n",
      "======================\n",
      "Wrote simplified GRG with:\n",
      "  Nodes: 15481\n",
      "  Edges: 93351\n",
      "Wrote GRG to igd_convert.example.grg\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# -j controls how many threads to use.\n",
    "grg construct -j 1 igd_convert.example.igd -o igd_convert.example.grg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2e34581-82c4-4e5f-aca6-ef218d78645e",
   "metadata": {},
   "source": [
    "## What if I have metadata?\n",
    "\n",
    "IGD and GRG have the general philosophy that a lot of metadata is kept separate from the genotype data. The metadata that the two formats contain natively are:\n",
    "* IGD: Individual identifiers (one per sampled individual, not haplotype) - see [pyigd.IGDReader.get_individual_ids](https://pyigd.readthedocs.io/en/latest/pyigd.html#pyigd.IGDReader.get_individual_ids).\n",
    "* IGD: Variant identifiers (one per variant) - see [pyigd.IGDReader.get_variant_ids](https://pyigd.readthedocs.io/en/latest/pyigd.html#pyigd.IGDReader.get_variant_ids).\n",
    "* GRG: Individual identifiers (same as IGD) - see [pygrgl.GRG.get_individual_id](https://grgl.readthedocs.io/en/stable/python_api.html#pygrgl.GRG.get_individual_id).\n",
    "\n",
    "Beyond that, it is suggested that you keep metadata in a simple format. For example, `igdtools` supports [exporting metadata to .txt files](https://picovcf.readthedocs.io/en/latest/igdtools.html#convert-vcf-gz-to-igd-and-export-metadata) in a format that is loadable by [numpy.loadtxt](https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html). This metadata can be accessed by the index of the variant (mutation), $i$, and you can also keep a mapping from variant identifier $v_i$ to $i$ so that you can easily lookup other metadata by the variant id (`igdtools` exports such a mapping)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adca4f9c-c5bb-4236-84ee-dbce4fcc15c1",
   "metadata": {},
   "source": [
    "## Related Topics\n",
    "\n",
    "* The [igdtools documentation](https://picovcf.readthedocs.io/en/latest/igdtools.html)\n",
    "* An overview of the [IGD file format](https://picovcf.readthedocs.io/en/latest/igd_overview.html)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
