{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97b7771a-1daf-4ec0-a26b-4e169e4332d6",
   "metadata": {},
   "source": [
    "# LinearOperators Tutorial\n",
    "\n",
    "Unlike some of the other tutorials, this one is focused on the Python API only, and a generic mechanism to integrate with iterative linear algebra methods from SciPy (or other libraries). The PCA and GWAS functions actually make use of the [LinearOperators](https://grapp.readthedocs.io/en/latest/grapp.html#non-standardized-linear-operators) under the hood.\n",
    "\n",
    "**What you'll need:**\n",
    "\n",
    "* Python dependencies \"grapp\", \"pandas\", \"seaborn\", \"scipy\": `pip install grapp pandas seaborn scipy`\n",
    "* Command line tool \"wget\": `sudo apt install wget` (or your system's equivalent)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "416aecde-633e-4e70-be90-4c68af0ccf17",
   "metadata": {},
   "source": [
    "## Get Dataset\n",
    "\n",
    "We're going to use a really small existing (simulated) dataset so that we can compare GRG to standard dense matrix algebra. As soon as datasets get even a little bit big, the dense matrix can take up too much RAM for demonstration purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9cc393c0-8113-4675-af50-59229dd61536",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2026-02-06 12:39:52--  https://github.com/aprilweilab/grg_pheno_sim/raw/refs/heads/main/demos/data/test-200-samples.vcf.gz\n",
      "Resolving github.com (github.com)... 140.82.112.4\n",
      "Connecting to github.com (github.com)|140.82.112.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://raw.githubusercontent.com/aprilweilab/grg_pheno_sim/refs/heads/main/demos/data/test-200-samples.vcf.gz [following]\n",
      "--2026-02-06 12:39:52--  https://raw.githubusercontent.com/aprilweilab/grg_pheno_sim/refs/heads/main/demos/data/test-200-samples.vcf.gz\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 494022 (482K) [application/octet-stream]\n",
      "Saving to: ‘linop.example.vcf.gz’\n",
      "\n",
      "     0K .......... .......... .......... .......... .......... 10% 2.07M 0s\n",
      "    50K .......... .......... .......... .......... .......... 20% 2.52M 0s\n",
      "   100K .......... .......... .......... .......... .......... 31% 4.32M 0s\n",
      "   150K .......... .......... .......... .......... .......... 41% 5.12M 0s\n",
      "   200K .......... .......... .......... .......... .......... 51% 2.94M 0s\n",
      "   250K .......... .......... .......... .......... .......... 62% 3.31M 0s\n",
      "   300K .......... .......... .......... .......... .......... 72%  119M 0s\n",
      "   350K .......... .......... .......... .......... .......... 82% 4.37M 0s\n",
      "   400K .......... .......... .......... .......... .......... 93% 3.36M 0s\n",
      "   450K .......... .......... .......... ..                   100% 6.98M=0.1s\n",
      "\n",
      "2026-02-06 12:39:53 (3.75 MB/s) - ‘linop.example.vcf.gz’ saved [494022/494022]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 10893 total variants\n",
      "Of which 6228 were written sparsely\n",
      "Stats for linop.example.igd\n",
      "... in range 0 - 18446744073709551615\n",
      "  Variants in range: 10893\n",
      "  Average samples/var: 50.2075\n",
      "  Stddev samples/var: 86.1985\n",
      "  Average var/sample: 1367.28\n",
      "  Stddev var/sample: 25.9211\n",
      "  Variants with missing data: 0\n",
      "  Total missing alleles: 0\n",
      "  Total unique sites: 10885\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if [[ ! -e linop.example.igd ]]; then\n",
    "  # Download a small example dataset\n",
    "  wget https://github.com/aprilweilab/grg_pheno_sim/raw/refs/heads/main/demos/data/test-200-samples.vcf.gz -O linop.example.vcf.gz\n",
    "\n",
    "  # Convert to IGD; this isn't necessary, but most of the time you will want to do this\n",
    "  igdtools linop.example.vcf.gz -o linop.example.igd\n",
    "fi\n",
    "\n",
    "# Just show some stats about the dataset\n",
    "igdtools -s linop.example.igd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad7d2c8f-0230-4e24-bc7c-fc6ac5711dbf",
   "metadata": {},
   "source": [
    "Now construct the GRG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4baec85-786c-4f9b-a02d-06dbd6cec540",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing input file in 85 parts.\n",
      "Auto-calculating number of trees per part.\n",
      "Converting segments of input data to graphs\n",
      "100%|██████████| 85/85 [00:00<00:00, 134.66it/s]\n",
      "Merging...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== GRG Statistics ===\n",
      "Nodes: 15481\n",
      "Edges: 93351\n",
      "Samples: 400\n",
      "Mutations: 10893\n",
      "Ploidy: 2\n",
      "Phased: true\n",
      "Populations: 0\n",
      "Range of mutations: 55829 - 9999127\n",
      "Specified range: 0 - 10894\n",
      "======================\n",
      "Wrote simplified GRG with:\n",
      "  Nodes: 15481\n",
      "  Edges: 93351\n",
      "Wrote GRG to linop.example.grg\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if [[ ! -e linop.example.grg ]]; then\n",
    "  # -j controls how many threads to use.\n",
    "  grg construct -j 1 linop.example.igd -o linop.example.grg\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "911c2776-e092-42f4-8069-df7fe0c4139b",
   "metadata": {},
   "source": [
    "The first thing we'll do is define a function that _very inneficiently_ converts a GRG to a dense matrix. This is only valuable for illustration and testing purposes, as these matrices will get huge while a GRG will remain small enough to fit in RAM still.\n",
    "For reasons of clarity, we're only going to generate the diploid genotype matrix (0, 1, 2 values) and we're going to pretend there is no such thing as missing data. For a fuller implementation that handles these things, see [testing_utils.py](https://github.com/aprilweilab/grapp/blob/main/test/testing_utils.py)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d3f165ab-e01f-4664-bde9-c953608bfad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pygrgl\n",
    "import numpy\n",
    "\n",
    "def grgToDiploidX(grg: pygrgl.GRG):\n",
    "    \"\"\"\n",
    "    Create a diploid genotype matrix from a GRG.\n",
    "    \"\"\"\n",
    "    result = numpy.zeros((grg.num_individuals, grg.num_mutations))\n",
    "    samps_below = [list() for _ in range(grg.num_nodes)]\n",
    "    for node_id in range(grg.num_nodes):\n",
    "        sb = []\n",
    "        if grg.is_sample(node_id):\n",
    "            sb.append(node_id)\n",
    "        for child_id in grg.get_down_edges(node_id):\n",
    "            sb.extend(samps_below[child_id])\n",
    "        samps_below[node_id] = sb\n",
    "\n",
    "        muts = grg.get_mutations_for_node(node_id)\n",
    "        if muts:\n",
    "            for sample_id in sb:\n",
    "                indiv = sample_id // grg.ploidy\n",
    "                for mut_id in muts:\n",
    "                    result[indiv][mut_id] += 1\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2d936365-4fa8-4678-ab96-ee4077be901a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the GRG dataset\n",
    "grg = pygrgl.load_immutable_grg(\"linop.example.grg\", load_up_edges=False)\n",
    "\n",
    "# Convert GRG to dense matrix\n",
    "X = grgToDiploidX(grg)\n",
    "\n",
    "assert X.shape == (grg.num_individuals, grg.num_mutations)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccece086-9122-4b75-b9fe-24a80c6389ee",
   "metadata": {},
   "source": [
    "The simplest thing to start with is just a random matrix multiplication against the dense matrix and the [grapp.linalg.ops_scipy.SciPyXOperator](https://grapp.readthedocs.io/en/latest/grapp.html#grapp.linalg.ops_scipy.SciPyXOperator). This operator, and the others in `grapp.linalg.ops_scipy` conform to the `scipy` [LinearOperator interface](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html#scipy.sparse.linalg.LinearOperator), which means that `_matmat()`, `_rmatmat()`, `_matvec()`, and `_rmatvec()` implement multiplications between the GRG and a matrix or vector."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8102bfe7-1d3e-43f2-9bb2-4cb745ae21ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from grapp.linalg.ops_scipy import SciPyXOperator\n",
    "\n",
    "# Perform a multiplication against X (NxM matrix, where N is number of individuals and M is number of mutations).\n",
    "random_mat = numpy.random.standard_normal((7, grg.num_individuals))\n",
    "\n",
    "# Dense matrix multiplication\n",
    "numpy_result = random_mat @ X\n",
    "\n",
    "# GRG matrix multiplication\n",
    "grg_X = SciPyXOperator(grg, pygrgl.TraversalDirection.UP)\n",
    "grg_result = random_mat @ grg_X\n",
    "\n",
    "numpy.allclose(numpy_result, grg_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "076c07d4-df4c-4150-bac5-cbe34b431115",
   "metadata": {},
   "source": [
    "That very simple result is equivalent to using the [pygrgl.matmul](https://grgl.readthedocs.io/en/stable/python_api.html#pygrgl.matmul) function with `by_individual=True`, it just has the nice feature of supporting the `numpy` matrix multiplication operator `@`, and also can be used with `scipy`'s linear algebra methods.\n",
    "\n",
    "You can also do matrix-vector products, or products with the transpose, using `numpy` syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d4074869-387d-438b-bfd2-7c6bbf3b661c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vector result has shape (10893,)\n",
      "Transpose result has shape (3, 200)\n"
     ]
    }
   ],
   "source": [
    "vect_result = random_mat[2] @ grg_X  # Vector x Matrix\n",
    "print(f\"Vector result has shape {vect_result.shape}\")\n",
    "\n",
    "B = numpy.random.standard_normal((3, grg.num_mutations))\n",
    "transpose_result = B @ grg_X.T  # B x X^T\n",
    "print(f\"Transpose result has shape {transpose_result.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "602daa85-e5d1-45d8-91e0-d13e4cd8222a",
   "metadata": {},
   "source": [
    "The real power of GRG LinearOperators comes from the transformations they can make to the genotype matrix. Here we list some out:\n",
    "* [grapp.linalg.ops_scipy.SciPyStdXOperator](https://grapp.readthedocs.io/en/latest/grapp.html#grapp.linalg.ops_scipy.SciPyStdXOperator): Operations against the standardized genotype matrix.\n",
    "* [grapp.linalg.ops_scipy.SciPyStdXTXOperator](https://grapp.readthedocs.io/en/latest/grapp.html#grapp.linalg.ops_scipy.SciPyStdXTXOperator): Operations against the matrix $X^T \\times X$ for standardized $X$, which is the correlation matrix of the genotype matrix.\n",
    "* [grapp.linalg.ops_scipy.SciPyStdXXTOperator](https://grapp.readthedocs.io/en/latest/grapp.html#grapp.linalg.ops_scipy.SciPyStdXXTOperator): Operations against the matrix $X \\times X^T$ for standardized $X$, which is the genetic relatedness matrix (GRM).\n",
    "\n",
    "Below we illustrate how these can interact with `scipy`, by getting the first $k$ eigenvalues from the GRM using [scipy.sparse.linalg.eigs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigs.html#scipy.sparse.linalg.eigs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86ff4b8d-fbfc-4aee-8395-425627369e26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dense result: [31851.93496578+0.j 28802.06816356+0.j 25479.32804366+0.j\n",
      " 23571.51144699+0.j 21472.83667251+0.j]\n",
      "GRG result: [31851.93496578+0.j 28802.06816356+0.j 25479.32804366+0.j\n",
      " 23571.51144699+0.j 21472.83667251+0.j]\n"
     ]
    }
   ],
   "source": [
    "from grapp.linalg.ops_scipy import SciPyStdXXTOperator\n",
    "from scipy.sparse.linalg import eigs\n",
    "from grapp.util import allele_frequencies\n",
    "\n",
    "# We need the allele frequencies for standardizing the genotype matrix\n",
    "freqs = allele_frequencies(grg)\n",
    "\n",
    "# Let's standardize the dense matrix X first, by column\n",
    "Xstd = numpy.zeros(X.shape)\n",
    "sigma = numpy.sqrt(2 * freqs * (1 - freqs))\n",
    "Xstd = numpy.divide(X - 2*freqs, sigma, out=Xstd, where=(sigma != 0))\n",
    "\n",
    "# First 5 eigen values of the dense GRM\n",
    "np_eigvals, _ = eigs(Xstd @ Xstd.T, k=5)\n",
    "\n",
    "# Now we can implicitly get the standardized GRM from the GRG by using a LinearOperator\n",
    "grg_GRM = SciPyStdXXTOperator(grg, freqs)\n",
    "grg_eigvals, _ = eigs(grg_GRM, k=5)\n",
    "\n",
    "assert numpy.allclose(np_eigvals, grg_eigvals)\n",
    "print(f\"Dense result: {np_eigvals}\")\n",
    "print(f\"GRG result: {np_eigvals}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adca4f9c-c5bb-4236-84ee-dbce4fcc15c1",
   "metadata": {},
   "source": [
    "## Related Topics\n",
    "\n",
    "* [PCA](PCA.html) is an example functionality that uses LinearOperators under the hood.\n",
    "* Documentation links:\n",
    "  * [grapp.linalg](https://grapp.readthedocs.io/en/latest/grapp.html#linear-algebra): Python APIs for Linear Algebra"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
